mani_skill/envs/tasks/digital_twins/push_T.py

```diff
@@ -0,0 +1,509 @@
+                    builder.add_box_collision(pose=second_block_pose, half_size=second_block_size,material=tee_material)
+                    #builder.add_box_collision(pose=second_block_pose, half_size=second_block_size)
+                builder.add_box_visual(pose=second_block_pose, half_size=second_block_size, material=sapien.render.RenderMaterial(
+                    base_color=base_color,
+                ),)
+                if not target:
+                    return builder.build(name=name)
+                else: return builder.build_kinematic(name=name)
+
+        self.tee = create_tee(name="Tee", target=False)
+        self.goal_tee = create_tee(name="goal_Tee", target=True, base_color=np.array([128,128,128,255])/255)
+
+        ##adding end-effector end-episode goal position
+        builder = self.scene.create_actor_builder()
+        builder.add_cylinder_visual(
+            radius=0.02,
+            half_length=1e-4,
+            material=sapien.render.RenderMaterial(base_color=np.array([128, 128, 128, 255]) / 255),
+        )
+        self.ee_goal_pos = builder.build_kinematic(name="goal_ee")
+
+        ############## Rest of function is Custom 2D "Pseudo-Rendering" function for fully batched GPU T Intersection Area Calculation ##############
+        res = 64
+        uv_half_width = 0.15
+        self.uv_half_width = uv_half_width
+        self.res = res
+        oned_grid = (torch.arange(res, dtype=torch.float32).view(1,res).repeat(res,1) - (res/2))
+        self.uv_grid = (torch.cat([oned_grid.unsqueeze(0), (-1*oned_grid.T).unsqueeze(0)], dim=0) + 0.5) / ((res/2)/uv_half_width)
+        self.uv_grid = self.uv_grid.to(self.device)
+        self.homo_uv = torch.cat([self.uv_grid, torch.ones_like(self.uv_grid[0]).unsqueeze(0)], dim=0)
+        
+        ##tee render
+        ## tee is made of two different boxes, and then translated by center of mass
+        self.center_of_mass = (0,0.0375) #in frame of upside tee with center of horizontal box (add cetner of mass to get to real tee frame)
+        box1 = torch.tensor([[-0.1, 0.025], [0.1, 0.025], [-0.1, -0.025], [0.1, -0.025]]) 
+        box2 = torch.tensor([[-0.025, 0.175], [0.025, 0.175], [-0.025, 0.025], [0.025, 0.025]])
+        box1[:, 1] -= self.center_of_mass[1]
+        box2[:, 1] -= self.center_of_mass[1]
+
+        #convert tee boxes to indices
+        box1 *= ((res/2)/uv_half_width)
+        box1 += (res/2)
+
+        box2 *= ((res/2)/uv_half_width)
+        box2 += (res/2)
+
+        box1 = box1.long()
+        box2 = box2.long()
+
+        self.tee_render = torch.zeros(res,res)
+        ##image map has flipped x and y, set values in transpose to undo
+        self.tee_render.T[box1[0,0]:box1[1,0], box1[2,1]:box1[0,1]] = 1
+        self.tee_render.T[box2[0,0]:box2[1,0], box2[2,1]:box2[0,1]] = 1
+        #image map y is flipped of xy plane, flip to unflip
+        self.tee_render = self.tee_render.flip(0).to(self.device)
+
+        def quat_to_z_euler(quats):
```

Comment:
these functions (i did not verify) may be useful to be part of the standard geometry submodule in maniskill in the future. Can leave it here for now though.